drivers/efuse/efuse.c

```diff
@@ -0,0 +1,395 @@
+ *   the "upper half" efuse device and registers that device so that can
+ *   be used by application code.
+ *
+ * Input Parameters:
+ *   dev path - The full path to the driver to be registered in the NuttX
+ *     pseudo-filesystem.  The recommended convention is to name all
+ *     efuse drivers as "/dev/efuse".
+ *   lower - A pointer to an instance of lower half efuse driver.  This
+ *     instance is bound to the efuse driver and must persists as long as
+ *     the driver persists.
+ *
+ * Returned Value:
+ *   On success, a non-NULL handle is returned to the caller.  In the event
+ *   of any failure, a NULL value is returned.
+ *
+ ****************************************************************************/
+
+FAR void *efuse_register(FAR const char *path,
+                         FAR struct efuse_lowerhalf_s *lower)
+{
+  FAR struct efuse_upperhalf_s *upper;
+  int ret;
+
+  DEBUGASSERT(path && lower);
+  minfo("Entry: path=%s\n", path);
+
+  /* Allocate the upper-half data structure */
+
+  upper = (FAR struct efuse_upperhalf_s *)
+          kmm_zalloc(sizeof(struct efuse_upperhalf_s));
+  if (!upper)
+    {
+      merr("Upper half allocation failed\n");
+      goto errout;
+    }
+
+  /* Initialize the efuse timer device structure (it was already zeroed
+   * by kmm_zalloc()).
+   */
+
+  nxsem_init(&upper->exclsem, 0, 1);
+  upper->lower = lower;
+
+  /* Copy the registration path */
+
+  upper->path = strdup(path);
+  if (!upper->path)
+    {
+      merr("Path allocation failed\n");
+      goto errout_with_upper;
+    }
+
+  /* Register the efuse timer device */
+
+  ret = register_driver(path, &g_efuseops, 0666, upper);
+  if (ret < 0)
+    {
+      merr("register_driver failed: %d\n", ret);
+      goto errout_with_path;
+    }
+
+  return (FAR void *)upper;
+
+errout_with_path:
+  kmm_free(upper->path);
+
+errout_with_upper:
+  nxsem_destroy(&upper->exclsem);
+  kmm_free(upper);
+
+errout:
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: efuse_unregister
+ *
+ * Description:
+ *   This function can be called to disable and unregister the efuse
+ *   device driver.
+ *
+ * Input Parameters:
+ *   handle - This is the handle that was returned by efuse_register()
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void efuse_unregister(FAR void *handle)
+{
+  FAR struct efuse_upperhalf_s *upper;
+  FAR struct efuse_lowerhalf_s *lower;
+
+  /* Recover the pointer to the upper-half driver state */
+
+  upper = (FAR struct efuse_upperhalf_s *)handle;
+  DEBUGASSERT(upper != NULL);
+  lower = upper->lower;
+  DEBUGASSERT(lower != NULL);
+
+  minfo("Unregistering: %s\n", upper->path);
+
+  /* Unregister the efuse timer device */
+
+  unregister_driver(upper->path);
+
+  /* Then free all of the driver resources */
+
+  kmm_free(upper->path);
+  nxsem_destroy(&upper->exclsem);
+  kmm_free(lower);
```

Comment:
should we let caller free lower?

Reply:
Yes, please look the watchdog_unregister() at drivers/timers/watchdog.c

Reply:
But I can't find kmm_free(lower) in watchdog_unregister:
```
void watchdog_unregister(FAR void *handle)
{
  FAR struct watchdog_upperhalf_s *upper;
  FAR struct watchdog_lowerhalf_s *lower;

  /* Recover the pointer to the upper-half driver state */

  upper = (FAR struct watchdog_upperhalf_s *)handle;
  DEBUGASSERT(upper != NULL);
  lower = upper->lower;
  DEBUGASSERT(lower != NULL);

  wdinfo("Unregistering: %s\n", upper->path);

#ifdef CONFIG_WATCHDOG_AUTOMONITOR
  watchdog_automonitor_stop(upper);
#endif

  /* Disable the watchdog timer */

  DEBUGASSERT(lower->ops->stop); /* Required */
  lower->ops->stop(lower);

  /* Unregister the watchdog timer device */

  unregister_driver(upper->path);

  /* Then free all of the driver resources */

  kmm_free(upper->path);
  nxsem_destroy(&upper->exclsem);
  kmm_free(upper);
}
```

Reply:
You are right @xiaoxiang781216 I will fix it! Thank you!